From cf731e6bb9043b9d3400f102e43c7b9e5325a310 Mon Sep 17 00:00:00 2001
From: vamsikrishnak <vamsikrishnak@ami.com>
Date: Thu, 9 Nov 2023 00:38:29 +0530
Subject: [PATCH] Fixed Task Monitor response after Taskcompleted

Description:

In Task Service, task monitor URI, while task is ongoing, On GET request
"redfish/v1/TaskService/Tasks/{{task_id}}/Monitor" it will return response
code as 202 and the current state of task as response. But after task completion ,
On GET request  it will return 204 the first time and then on it will return
404 Not Found.

Expected 404 Not Found after task completion  and the attribute task monitor
should not be listed in task instance after its completion.

The commit fixes the above issue by changing code in task.hpp.

Testing:
  Tested,Monitor the task and check the response after completing the task.

Signed-off-by: vamsikrishnak <vamsikrishnak@ami.com>
---
 redfish-core/lib/task.hpp | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/redfish-core/lib/task.hpp b/redfish-core/lib/task.hpp
index 50de31d..2934461 100644
--- a/redfish-core/lib/task.hpp
+++ b/redfish-core/lib/task.hpp
@@ -160,10 +160,9 @@ struct TaskData : std::enable_shared_from_this<TaskData>
             res.addHeader(boost::beast::http::field::retry_after,
                           std::to_string(retryAfterSeconds));
         }
-        else if (!gave204)
+        else if (!taskCompleted)
         {
-            res.result(boost::beast::http::status::no_content);
-            gave204 = true;
+            taskCompleted = true;
         }
     }
 
@@ -320,7 +319,7 @@ struct TaskData : std::enable_shared_from_this<TaskData>
     std::unique_ptr<sdbusplus::bus::match_t> match;
     std::optional<time_t> endTime;
     std::optional<Payload> payload;
-    bool gave204 = false;
+    bool taskCompleted = false;
     int percentComplete = 0;
 };
 
@@ -415,13 +414,13 @@ inline void requestRoutesTaskMonitor(App& app)
             return;
         }
         std::shared_ptr<task::TaskData>& ptr = *find;
-        // monitor expires after 204
-        if (ptr->gave204)
+        ptr->populateResp(asyncResp->res);
+        // monitor expires after taskCompleted
+        if (ptr->taskCompleted)
         {
             messages::resourceNotFound(asyncResp->res, "Task", strParam);
             return;
         }
-        ptr->populateResp(asyncResp->res);
     });
 }
 
@@ -473,7 +472,7 @@ inline void requestRoutesTask(App& app)
         asyncResp->res.jsonValue["Messages"] = ptr->messages;
         asyncResp->res.jsonValue["@odata.id"] =
             boost::urls::format("/redfish/v1/TaskService/Tasks/{}", strParam);
-        if (!ptr->gave204)
+        if (!ptr->taskCompleted)
         {
             asyncResp->res.jsonValue["TaskMonitor"] =
                 "/redfish/v1/TaskService/Tasks/" + strParam + "/Monitor";
-- 
2.9.5

